dkc/core/rest/file.py

```diff
@@ -31,12 +33,18 @@ class Meta:
 class FileViewSet(ModelViewSet):
     queryset = File.objects.all()
 
-    permission_classes = [HasAccess]
+    permission_classes = [HasAccess & IsAuthenticatedOrReadOnly]
     serializer_class = FileSerializer
 
     filter_backends = [PermissionFilterBackend, ActionSpecificFilterBackend]
     filterset_fields = ['folder', 'sha512']
 
+    def perform_create(self, serializer: FileSerializer):
+        folder: Folder = serializer.validated_data.get('folder')
+        if not folder.has_permission(self.request.user, permission=Permission.write):
+            raise PermissionDenied()
+        serializer.save()
```

Comment:
This contradicts what I said this morning, but I now realize this logic probably belongs in a permission class to be consistent with how the other permissions are enforced.  I'm thinking perhaps to add an explicit API for object creation on the model classes.

Something like this:
```python
class Cls:
	@classmethod
    def can_create(cls, user, obj) -> bool:
		# return true if the user has permission to create the object
```

Then, a generic permissions class that calls that method.  The thing I'm not sure about is how to create the object inside permission class.  I don't know if you can get the serializer from the request object or not...  Alternatively, we could pass `request.data` into that method.

If this gets too hairy, I think we can save it for later because it's an important bug fix.

Reply:
I'm all for doing this the right way, though I'd rather have something that works in place, and I don't think I'm personally familiar enough with the best practices here to make an attempt at your suggestion. Maybe we can merge this and open an issue to move it.

Reply:
Works for me.  I can make an attempt at it.